portafly/src/components/AppLayout.tsx

```diff
@@ -0,0 +1,172 @@
+import * as React from 'react'
+import {
+  Nav,
+  NavList,
+  NavVariants,
+  Page,
+  PageHeader,
+  PageSidebar,
+  SkipToContent,
+  PageHeaderProps,
+} from '@patternfly/react-core'
+import {AppNavExpandable, IAppNavExpandableProps} from './AppNavExpandable'
+import {AppNavGroup, IAppNavGroupProps} from './AppNavGroup'
+import {AppNavItem, IAppNavItemProps} from './AppNavItem'
+
+export interface IAppLayoutContext {
+  setBreadcrumb: (breadcrumb: React.ReactNode) => void
+}
+
+export const AppLayoutContext = React.createContext<IAppLayoutContext>({
+  setBreadcrumb: () => void 0,
+})
+
+export interface IAppLayoutProps
+  extends Pick<PageHeaderProps, 'logo'>,
+    Pick<PageHeaderProps, 'logoProps'>,
+    Pick<PageHeaderProps, 'avatar'>,
+    Pick<PageHeaderProps, 'toolbar'> {
+  navVariant?: 'vertical' | 'horizontal'
+  navItems?: Array<IAppNavItemProps | IAppNavExpandableProps | IAppNavGroupProps | undefined>
+  navGroupsStyle?: 'grouped' | 'expandable'
+  startWithOpenNav?: boolean
+  theme?: 'dark' | 'light'
+  mainContainerId?: string
+}
+
+export const AppLayout: React.FunctionComponent<IAppLayoutProps> = ({
+  logo,
+  logoProps,
+  navVariant = 'horizontal',
+  navItems = [],
+  navGroupsStyle = 'grouped',
+  toolbar,
+  avatar,
+  startWithOpenNav = true,
+  theme = 'dark',
+  mainContainerId = 'main-container',
+  children,
+}) => {
+  const [isNavOpen, setIsNavOpen] = React.useState(startWithOpenNav)
+  const [isMobileView, setIsMobileView] = React.useState(true)
+  const [isNavOpenMobile, setIsNavOpenMobile] = React.useState(false)
+  const [breadcrumb, setBreadcrumb] = React.useState<React.ReactNode | undefined>()
+  const previousBreadcrumb = React.useRef<React.ReactNode | null>()
```

Comment:
Why null and not undefined?

Reply:
From here until the end of your review, it's been imported (as the commit message said) from Fuse team's use-patternfly repo. Instead of using it as a dependency, we used it as a bootstrapping tool. Which allow us to make changes in order to adapt it to our needs... that said, I don't have all the answers

Reply:
But...both are ok, the definition of `useRef` accepts both cases. I think in this case it wants to keep the type of object instead of undefined.